src/geo/transform.js

```diff
@@ -1329,6 +1335,18 @@ class Transform {
         return cache[projMatrixKey];
     }
 
+    calculatePixelsToTileUnitsMatrix(tile: Tile): Float32Array {
+        const key = tile.tileID.key;
+        const cache = this._pixelsToTileUnitsCache;
+        if (cache[key]) {
+            return cache[key];
+        }
+
+        const matrix = getPixelsToTileUnitsMatrix(tile, this);
+        cache[key] = matrix;
+        return cache[key];
+    }
```

Comment:
It seems like the only expensive part of the `getPixelsToTileUnitsMatrix` calculation is getting the `tileTransform` for the tile/projection pair, but it is 1) not dependent on fractional zoom, 2) used and redundantly recalculated in a ton of other places. 3) will get more expensive with my `projections-constrain` branch changes that make it more precise.

So I think that instead of caching the units matrix, we should cache the tile transform and make sure it's used in other places — this should both reduce the overhead of this change and elsewhere in the projections code.

Reply:
I removed the `tileTransform` recalculation by storing it on the tile objects. I left the matrix caching because quick profiling showed it slightly faster and it's a tiny bit more convenient to get them through `Transform`